### ISSUE TITLE
Issue with long audio files and streaming models

### ISSUE THREAD

**Issue by espnetUser** [espnet/espnet#4217]
Thanks for a great toolkit! 

I am working on ESPNET2 Streaming Conformer model (based on espnet2 Librispeech setup) adapted to my own data set and get very good results for non-streaming model. For evaluating non-streaming model I split up my test set of long audio files (> 20min) into smaller segments (<15s). 

Now with the streaming model my assumption was I can directly work with longer audio files as processing is done on blocks rather than entire speech signal. But I still experience OOM issues during cpu-based decoding even on machines with 245G of RAM.

I will implement an internal VAD to the decoding loop as suggested in #3368 but would like to understand if there are some fundamental limitations in current espnet implementation that prevent use in real streaming scenario.

So my questions are:

1. I noticed that the OOM issues reduced somehwat during streaming inference when disabling the repitition detection via "disable_repetition_detection: true". The following code comment in espnet/nets/batch_beam_search_online.py about repetition detection:
```
# NOTE(tsunoo): check repetitions here
# This is a implicit implementation of
# Eq (11) in https://arxiv.org/abs/2006.14941
# A flag prev_repeat is used instead of using set
# NOTE(fujihara): I made it possible to turned off
# the below lines using disable_repetition_detection flag,
# because this criteria is too sensitive that the beam
# search starts only after the entire inputs are available.
# Empirically, this flag didn't affect the performance.
```
seems to indicate that 
1) repetion detection is not suitable for streaming application as it requires entire inputs available before beam search starts 
2) disabling it has no adverse impact on performance
	
That seems to contradict https://arxiv.org/abs/2006.14941 where repetition detection was found to be an important part of the streaming E2E model and if not used caused significant increase in word error rates.
	
Is my understanding correct or did I miss something here?

2. I also saw the last comment in #3368 that says there are still issues with long audio recordings in case of streaming and noticed the docstring for BatchBeamSearchOnline class:
```
class BatchBeamSearchOnline(BatchBeamSearch):
    """Online beam search implementation.

    This simulates streaming decoding.
    It requires encoded features of entire utterance and
    extracts block by block from it as it shoud be done
    in streaming processing.
    This is based on Tsunoo et al, "STREAMING TRANSFORMER ASR
    WITH BLOCKWISE SYNCHRONOUS BEAM SEARCH"
    (https://arxiv.org/abs/2006.14941).
    """
```
@sw005320: Could you please provide more details on what exactly these issues are and why BatchBeamSearchOnline still requires the encoded features of the entire utterance in online mode? 
Is it because of the repetition detection or something else?

Thank you!

**Comment by sw005320** [comment]
@D-Keqi, can you answer it for me?
About the BatchBeamSearchOnline, Is this comment true?
I think it is only for BatchBeamSearchOnlineSim

**Comment by D-Keqi** [comment]
> @D-Keqi, can you answer it for me? About the BatchBeamSearchOnline, Is this comment true? I think it is only for BatchBeamSearchOnlineSim

Sure!

**Comment by D-Keqi** [comment]
> Thanks for a great toolkit!
> 
> I am working on ESPNET2 Streaming Conformer model (based on espnet2 Librispeech setup) adapted to my own data set and get very good results for non-streaming model. For evaluating non-streaming model I split up my test set of long audio files (> 20min) into smaller segments (<15s).
> 
> Now with the streaming model my assumption was I can directly work with longer audio files as processing is done on blocks rather than entire speech signal. But I still experience OOM issues during cpu-based decoding even on machines with 245G of RAM.
> 
> I will implement an internal VAD to the decoding loop as suggested in #3368 but would like to understand if there are some fundamental limitations in current espnet implementation that prevent use in real streaming scenario.
> 
> So my questions are:
> 
> 1. I noticed that the OOM issues reduced somehwat during streaming inference when disabling the repitition detection via "disable_repetition_detection: true". The following code comment in espnet/nets/batch_beam_search_online.py about repetition detection:
> 
> ```
> # NOTE(tsunoo): check repetitions here
> # This is a implicit implementation of
> # Eq (11) in https://arxiv.org/abs/2006.14941
> # A flag prev_repeat is used instead of using set
> # NOTE(fujihara): I made it possible to turned off
> # the below lines using disable_repetition_detection flag,
> # because this criteria is too sensitive that the beam
> # search starts only after the entire inputs are available.
> # Empirically, this flag didn't affect the performance.
> ```
> 
> seems to indicate that
> 
> 1. repetion detection is not suitable for streaming application as it requires entire inputs available before beam search starts
> 2. disabling it has no adverse impact on performance
> 
> That seems to contradict https://arxiv.org/abs/2006.14941 where repetition detection was found to be an important part of the streaming E2E model and if not used caused significant increase in word error rates.
> 
> Is my understanding correct or did I miss something here?
> 
> 2. I also saw the last comment in [DefaultCPUAllocator: can't allocate memory #3368](https://github.com/espnet/espnet/issues/3368) that says there are still issues with long audio recordings in case of streaming and noticed the docstring for BatchBeamSearchOnline class:
> 
> ```
> class BatchBeamSearchOnline(BatchBeamSearch):
>     """Online beam search implementation.
> 
>     This simulates streaming decoding.
>     It requires encoded features of entire utterance and
>     extracts block by block from it as it shoud be done
>     in streaming processing.
>     This is based on Tsunoo et al, "STREAMING TRANSFORMER ASR
>     WITH BLOCKWISE SYNCHRONOUS BEAM SEARCH"
>     (https://arxiv.org/abs/2006.14941).
>     """
> ```
> 
> @sw005320: Could you please provide more details on what exactly these issues are and why BatchBeamSearchOnline still requires the encoded features of the entire utterance in online mode? Is it because of the repetition detection or something else?
> 
> Thank you!

@espnetUser Hi! BatchBeamSearchOnline does not require the entire utterance. Can you show me your configuration of streaming decoding? If you do not specify the value of "sim_chunk_length", the decoding is actually non-streaming, which you can refer to here https://github.com/espnet/espnet/blob/master/espnet2/bin/asr_inference_streaming.py#L411. If  the sim_chunk_length has been specified, I think it should be real streaming, and you can refer to here https://github.com/espnet/espnet/blob/master/espnet2/bin/asr_inference_streaming.py#L419

I hope this helps.

**Comment by espnetUser** [comment]
@D-Keqi: Thanks for your helpful reply! I  just had time to come back to this issue.

> Can you show me your configuration of streaming decoding?
```
$ cat conf/tuning/decode_asr_streaming.yaml
lm_weight: 0.0
ctc_weight: 0.4
beam_size: 20
penalty: 0.0
maxlenratio: 0.0
minlenratio: 0.0
ctc_weight: 0.4
lm_weight: 0.0
sim_chunk_length: 512
disable_repetition_detection: true
decoder_text_length_limit: 0
encoded_feat_length_limit: 0
```
The configuration looks fine to me and from the logfile I can see the decoder results get output in a streaming fashion.

> So if your audio is very long, this may cause you into OOM issue, but this model is still real streaming.

Thanks for pointing this out. This may explain the OOM issue.

**Comment by espnetUser** [comment]
@D-Keqi and @sw005320: Could you please clarify the following questions regarding streaming:

**1. Impact of repetition detection** 
In https://arxiv.org/abs/2006.14941 repetition detection was found to be an important part of the streaming E2E model and if not used caused significant increase in word error rates for CSJ and Librispeech tasks. But the repetition detection seems to be turned off by default in streaming mode configs
```
disable_repetition_detection: true
```
and the comment below suggests it cannot be be used in real streaming?
```
# NOTE(fujihara): I made it possible to turned off
# the below lines using disable_repetition_detection flag,
# because this criteria is too sensitive that the beam
# search starts only after the entire inputs are available.
# Empirically, this flag didn't affect the performance. 
```

**2. Long audio file issues**
I noticed that the beam search slows down dramatically for long audio files (e.g. 20min). Is this a limitation of the current implementation of the "blockwise synchronous beam search" and can be addressed in the future or is this a fundamental problem of the algorithm, for instance related to Eq. (14) in https://arxiv.org/abs/2006.14941?

**3. Workaround**
Is the use of ```encoded_feat_length_limit``` and ```decoder_text_length_limit``` intended to help with long input sequences or do they serve different purpose?

**Comment by D-Keqi** [comment]
> 



> @D-Keqi and @sw005320: Could you please clarify the following questions regarding streaming:
> 
> **1. Impact of repetition detection** In https://arxiv.org/abs/2006.14941 repetition detection was found to be an important part of the streaming E2E model and if not used caused significant increase in word error rates for CSJ and Librispeech tasks. But the repetition detection seems to be turned off by default in streaming mode configs
> 
> ```
> disable_repetition_detection: true
> ```
> 
> and the comment below suggests it cannot be be used in real streaming?
> 
> ```
> # NOTE(fujihara): I made it possible to turned off
> # the below lines using disable_repetition_detection flag,
> # because this criteria is too sensitive that the beam
> # search starts only after the entire inputs are available.
> # Empirically, this flag didn't affect the performance. 
> ```
> 
> **2. Long audio file issues** I noticed that the beam search slows down dramatically for long audio files (e.g. 20min). Is this a limitation of the current implementation of the "blockwise synchronous beam search" and can be addressed in the future or is this a fundamental problem of the algorithm, for instance related to Eq. (14) in https://arxiv.org/abs/2006.14941?
> 
> **3. Workaround** Is the use of `encoded_feat_length_limit` and `decoder_text_length_limit` intended to help with long input sequences or do they serve different purpose?

Hi, 
1. Sorry, actually I do not have too much knowledge about the repetition detection mechanism.
2. You have mentioned "slow down dramatically", So you mean  the decoding speed becomes slower or the degradation of ASR accuracy?

**Comment by sw005320** [comment]
@eml914, can you answer the rest of the questions for us?

**Comment by eml914** [comment]
Appologies for belated response.
The repetition detection sometimes becomes problematic, and I made some improvement in https://arxiv.org/abs/2201.10190 but not implemented yet.
The current streaming decoding cannot afford very long audio because the decoder still uses entire history whilst the encoder goes blockwise.  That is still future work and I have an idea to do sort of VAD within the streaming decoding, but not evaluated yet.
I am sorry for the inconvenience.

**Comment by eml914** [comment]
So, for now an option is that to use LSTM decoder instead of transformer, which would not use that much memory I believe.

**Comment by espnetUser** [comment]
Thank you @eml914 for your helpful response and my apologies for the late reply! 

> The current streaming decoding cannot afford very long audio because the decoder still uses entire history whilst the encoder goes blockwise. 

For now, I worked around the memory issue by splitting long audio files into smaller chunks using external VAD and glueing the hypos back together after decoding.

> That is still future work and I have an idea to do sort of VAD within the streaming decoding, but not evaluated yet.

Looking forward to trying it! :) A quick search found this 
"VAD-free Streaming Hybrid CTC/Attention ASR for Unsegmented Recording" (https://arxiv.org/pdf/2107.07509.pdf) which also seems to go in that direction to address the long audio issue.

> So, for now an option is that to use LSTM decoder instead of transformer, which would not use that much memory I believe.

Thanks, I will look into it. 
Do you mean a Conformer-Transducer setup like this:
https://github.com/espnet/espnet/blob/master/egs2/librispeech/asr1/conf/train_rnnt_conformer.yaml ?

**Comment by espnetUser** [comment]
One last question where I would really appreciate your comment. 

With respect to the contextual_block_conformer could you please clarify which of these parameters in the streaming config affects the "built-in" or "algorithmic" delay of the encoder: 
```
block_size: 40      # streaming configuration
hop_size: 16        # streaming configuration
look_ahead: 8       # streaming configuration
```
 
Based on   https://arxiv.org/abs/2006.14941 (page 6: _"The theoretical delay was 0.64 seconds because the encoder block shifted every 16 frames with 4-factor downsampling."_) I assume it was calculated like so:
```
built-in latency = hop_size * down_sample_factor * frame_shift 
                 = 16 * 4 * 10ms
                 = 640ms
```
I was wondering why the algorithmic latency is not controlled by the look_ahead value (plus maybe the center frames of the block, like in  https://arxiv.org/pdf/2010.10759.pdf, Section 3.2.1. Algorithmic latency induced by the encoder) but by the hop_size (shift) instead? 

Thanks!

**Comment by eml914** [comment]
@espnetUser Thank you for your comments and questions!
We have tried rnn-t training for several times but not yet get better result than usual transformer.  But for the long audio case, it might be the solution.  What I ment was using LSTM which does not use all the history (but it uses all the encoder output if you use attention mechanism) and might be good for the long audio.  But further investigation is required.

Regarding the theoretical delay, you are right and the paper is misleading. We have to take the look_ahead value into account.  So in the worst case, the delay would be
`latency = (hop_size + look_ahead) * down_sample_factor * frame_shift = 960ms`
But the latency depends on the timing of the encoder shift and the shortest case would be that the utterance just ends on the shifting timing and only look_ahead value contribute to the delay.
So, on average, the latency would be
`latency = (hop_size/2+look_ahead) * down_sample_factor * frame_shift = 640ms`
I should rewrite that part in the paper.  Thank you for pointing out.

**Comment by espnetUser** [comment]
@eml914: Thank you very much for the clarification!

**Comment by espnetUser** [comment]
@eml914: Sorry, to bother you again. Could you please help me understand the relationship between the following streaming config parameters of the block conformer encoder specified in the espnet configs
```
block_size: 40      # streaming configuration
hop_size: 16        # streaming configuration
look_ahead: 16      # streaming configuration
```
and the notation {N_l, N_c, N_r} for left/center/right frames of a block used in your paper https://arxiv.org/abs/2006.14941.
Is the following correct?
```
block_size = N_l + N_c + N_r
look_ahead = N_r
```
How would I specify N_l and N_c values to change the config from let's say {16, 16, 8} to {4, 8, 4}?

Thanks!

**Comment by eml914** [comment]
Yes, that is correct and  `hop_size = N_l` so
{4, 8, 4} would be
```
block_size: 16
hop_size: 4
look_ahead: 4
```

**Comment by espnetUser** [comment]
Ah, thanks! I somehow missed that ```hop_size = N_l```. 

My current config is this:
```
block_size: 40      # streaming configuration
hop_size: 16        # streaming configuration
look_ahead: 8      # streaming configuration
```
with downsample factor of 4 and 11.6 frame shift (128 sample shift @ 11025Hz).
Based on our previous discussion this gives me an average latency of:
```
avg_latency = (hop_size/2+look_ahead) * down_sample_factor * frame_shift 
                    = (16/2 + 8) * 4 * 11.6ms
                    = 742ms
```
I am aiming to reduce the average latency to around 300ms for my model. For example, I could achieve 371ms by using a downsampling factor of 2 instead of 4 (at cost of much higher compute). But would I also need to double the block length to 80 so the block encoder still operates on same context length to avoid increase in WER? Do you have any recommendations what works best to reduce latency without hurting WER?

**Comment by eml914** [comment]
I am very sorry.  I was confusing and the truth is `hop_size=N_c` so (4, 8, 4) would be
```
block_size: 16
hop_size: 8
look_ahead: 4
```
But anyway your configuration seems to be (16, 16, 8) and that sounds good for WER.
I initially used (16, 16, 0) but look ahead frames are important to reduce WER.
So I would recomend you to make it (24, 8, 8), or (32, 4, 4), which is quite extream and not computationally efficient at all.
```
block_size: 40
hop_size: 4
look_ahead: 4
```

**Comment by espnetUser** [comment]
Thanks @eml914 for your clarification and recommendations on the `hop_size=N_c` which of course makes sense. I am running some trainings with different configs which essentially decrease look-ahead and center context to lower latency/delay at cost of increasing left context size and see if I can find a streaming config with suitable latency for my use case.

One last question regarding the problem that lower latency configs tend to be not computationally efficient anymore. Do you think it would be possible for your contextual block encoder to implement a similar caching mechanism of the key and value in previous block's self-attention layers to avoid duplicated computations for left context blocks as proposed in https://arxiv.org/pdf/2010.10759.pdf?

**Comment by espnetUser** [comment]
Hi @eml914 and @sw005320, I just discovered with great interest your recent paper on Spiralformer for low latency streaming speech encoder (https://arxiv.org/abs/2510.00982). I think it addresses precicely the problem what we were discussing about in the last comments in this old issue from a couple of years ago ;). Could you please let me know if it is planned to release the code for Spiralformer in ESPNet or whether there is an implementation available to test this out? Thanks!

**Comment by sw005320** [comment]
Thanks for your interest. We will discuss it internally!

**Comment by espnetUser** [comment]
Hi @sw005320, I just wanted to check if a decision has been made yet regarding the release or implementation of the Spiralformer approach in ESPnet in the near future? Thanks!

**Comment by eml914** [comment]
Hi, I am sorry for my late response.
We can share the batched training script in ESPnet for Spiralformer, and you can also calcualte WER with the script. But, in the paper, we evaluate realtime inference with C++ program, which cannot be shared. Is that OK if we can only share the training script?

**Comment by espnetUser** [comment]
Thank you for sharing the batched training script in ESPnet for Spiralformer—that would be excellent! For real-time inference, I am planning to run the model in separate C++ code via ONNX. Does the Spiralformer architecture have any restrictions regarding ONNX export capability?

**Comment by eml914** [comment]
I have finally send a  Pull Request. https://github.com/espnet/espnet/pull/6510 
I am not fully sure the code works via ONNX but it is a basic transformer so should be. I only implemented by myself from scratch for the paper.
